rust/index/src/spann/utils.rs

```diff
@@ -0,0 +1,565 @@
+use std::{cmp::min, collections::HashMap};
+
+use chroma_distance::DistanceFunction;
+use rand::{seq::SliceRandom, Rng};
+
+// TODO(Sanket): I don't understand why the reference implementation defined
+// max_distance this way.
+const MAX_DISTANCE: f32 = f32::MAX / 10.0;
+const NUM_ITERS_FOR_CENTER_INIT: usize = 3;
+const NUM_ITERS_FOR_MAIN_LOOP: usize = 100;
+const NUM_ITERS_NO_IMPROVEMENT: usize = 5;
+
+struct KMeansAlgorithmInput<'referred_data> {
+    indices: Vec<u32>,
+    embeddings: &'referred_data [f32],
+    embedding_dimension: usize,
+    k: usize,
+    first: usize,
+    // Exclusive range.
+    last: usize,
+    num_samples: usize,
+    distance_function: DistanceFunction,
+    initial_lambda: f32,
+}
+
+#[allow(dead_code)]
+#[derive(Debug)]
+pub struct KMeansAlgorithmOutput {
+    cluster_centers: Vec<Vec<f32>>,
+    cluster_counts: Vec<usize>,
+    cluster_labels: HashMap<u32, i32>,
+}
+
+pub struct KMeansAlgorithm<'referred_data> {
+    input: KMeansAlgorithmInput<'referred_data>,
+}
+
+#[derive(Debug)]
+struct KMeansAssignForCenterInitOutput {
+    cluster_counts: Vec<usize>,
+    cluster_weighted_counts: Vec<f32>,
+    cluster_farthest_distance: Vec<f32>,
+    total_distance: f32,
+}
+
+#[allow(dead_code)]
+#[derive(Debug)]
+struct KMeansAssignForMainLoopOutput {
+    cluster_counts: Vec<usize>,
+    cluster_farthest_point_idx: Vec<i32>,
+    cluster_farthest_distance: Vec<f32>,
+    cluster_new_centers: Vec<Vec<f32>>,
+    total_distance: f32,
+}
+
+#[allow(dead_code)]
+#[derive(Debug)]
+struct KMeansAssignFinishOutput {
+    cluster_counts: Vec<usize>,
+    cluster_nearest_point_idx: Vec<i32>,
+    cluster_nearest_distance: Vec<f32>,
+    cluster_labels: HashMap<u32, i32>,
+}
+
+impl<'referred_data> KMeansAlgorithm<'referred_data> {
+    #[allow(clippy::too_many_arguments)]
+    pub fn new(
+        indices: Vec<u32>,
+        embeddings: &'referred_data [f32],
+        embedding_dimension: usize,
+        k: usize,
+        first: usize,
+        last: usize,
+        num_samples: usize,
+        distance_function: DistanceFunction,
+        initial_lambda: f32,
+    ) -> Self {
+        KMeansAlgorithm {
+            input: KMeansAlgorithmInput {
+                indices,
+                embeddings,
+                embedding_dimension,
+                k,
+                first,
+                last,
+                num_samples,
+                distance_function,
+                initial_lambda,
+            },
+        }
+    }
+
+    fn get_nearest_center(
+        &self,
+        centers: &[Vec<f32>],
+        idx: usize,
+        lambda: f32,
+        previous_counts: &[usize],
+    ) -> (i32, f32) {
+        let point_idx = self.input.indices[idx];
+        let dim = self.input.embedding_dimension;
```

Comment:
if we had helper structs for the embedding / dimension this would be made cleaner